### PR TITLE
[mypyc] Fix direct __dict__ access on inner functions in new Python

### LINKED ISSUE #16077: `functools.update_wrapper` and `functools.wraps` cannot be used in mypycified files on py312
# Bug Report

Mypycifying a file that uses `functools.update_wrapper` or `functools.wraps` breaks if you're using Python 3.12.

### To Reproduce

1. Ensure you're using Python 3.12
2. Save this code into a file `repro.py`:

   ```py
   from functools import update_wrapper
   from typing import Callable
   
   def dont_increase_indentation(split_func: Callable[[], str]) -> Callable[[], str]:
       def split_wrapper() -> str:
           return "foo"
  
       update_wrapper(split_wrapper, split_func)
       return split_wrapper


   @dont_increase_indentation
   def delimiter_split() -> str:
       return "bar"
   ```

4. Run `mypyc repro.py`
5. Run `python -c "import repro"`

### Expected Behavior

The mypycified `repro` module should import without errors. This is the behaviour you get if you mypycify the module using Python 3.11.

### Actual Behavior

```pytb
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "repro3.py", line 12, in <module>
    @dont_increase_indentation
  File "repro3.py", line 8, in dont_increase_indentation
    update_wrapper(split_wrapper, split_func)
  File "C:\Users\alexw\.pyenv\pyenv-win\versions\3.12.0rc2\Lib\functools.py", line 58, in update_wrapper
    getattr(wrapper, attr).update(getattr(wrapped, attr, {}))
    ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'split_wrapper_dont_increase_indentation_obj' object has no attribute '__dict__'. Did you mean: '__dir__'?
```

### What is causing the bug?

The bug appears to be due to mypyc not creating `__dict__` attributes for function objects. This appears to be despite this comment in the source code, which looks like it's intended to create `__dict__` attributes for functions precisely due to how `functools.update_wrapper` works:

https://github.com/python/mypy/blob/49419835045b09c98b545171abb10384b6ecf6a9/mypyc/irbuild/callable_class.py#L60-L65

### Why is the bug only occurring on Python 3.12?

The bug with `functools.update_wrapper` only occurs with Python 3.12. The interesting thing is that you can reproduce a similar bug with this snippet, but unlike the first snippet, this will _also_ repro the bug on Python 3.11:

```py
from typing import Callable

def dont_increase_indentation(split_func: Callable[[], str]) -> Callable[[], str]:
    def split_wrapper() -> str:
        return "foo"

    split_wrapper.__dict__.update(split_func.__dict__)
    return split_wrapper


@dont_increase_indentation
def delimiter_split() -> str:
    return "bar"
```

Error when running `python -c "import repro"` after mypycifying this snippet:

```pytb
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "repro2.py", line 12, in <module>
    @dont_increase_indentation
  File "repro2.py", line 8, in dont_increase_indentation
    split_wrapper.__dict__.update(split_func.__dict__)
AttributeError: 'split_wrapper_dont_increase_indentation_obj' object has no attribute '__dict__'. Did you mean: '__dir__'?
```

It's unclear to me why `functools.update_wrapper` works on Python 3.11 and breaks on Python 3.12 (snippet 1), when the "inlined" version (snippet 2) breaks on both Python 3.11 and Python 3.12.

### Did `functools.update_wrapper` change in Python 3.12?

There was a minor change to `update_wrapper` in Python 3.12 as part of the PEP-695 implementation: https://github.com/python/cpython/pull/104601. Other than this change, however, the source code for `update_wrapper` and `wraps` has remained unchanged for the last 10 years.

## Your Environment

- Mypy version used: 1.5.1
- Python version used: 3.12.0rc2
- Platform: Windows

### PR DESCRIPTION
**Pull request by hauntsaninja**
Fixes #16077

### CODE REVIEW

**Comment by AlexWaygood** [comment]
Nice! I can confirm that, with this patch, both snippets I pasted in #16077 work for me locally on Python 3.12 once they've been mypycified.

**Comment by AlexWaygood** [comment]
Hmm, looks like the new test passes with py312 but not with any other Python version?

**Comment by hauntsaninja** [comment]
Hmm maybe we can just drop support for all these legacy Python versions?

**Comment by hauntsaninja** [comment]
I opted for a simple "fix" for older Python versions, which is to test less. But I think we can make the `.__dict__` access actually work if we just remove `__dict__` from tp_members. All tests pass with that change, but it's possibly incorrect.

File: mypyc/test-data/run-functions.test [hunk 1]
```diff
@@ -1256,3 +1256,33 @@ def foo(**kwargs: Unpack[Person]) -> None:
 foo(name='Jennifer', age=38)
 [out]
 Jennifer
+
+[case testNestedFunctionDunderDict312]
+import sys
+
+def foo() -> None:
+    def inner() -> str: return "bar"
+    print(inner.__dict__)  # type: ignore[attr-defined]
```
Comment (JelleZijlstra) [inline mypyc/test-data/run-functions.test:1265]: There's a lot of type ignores in this test, do we really need that?

Comment (hauntsaninja) [reply mypyc/test-data/run-functions.test:1265]: yes, fixtures

File: mypyc/test-data/run-functions.test [hunk 2]
```diff
@@ -1256,3 +1256,33 @@ def foo(**kwargs: Unpack[Person]) -> None:
...
+{'x': 1}
+1
+
+[case testFunctoolsUpdateWrapper]
+import functools
+
+def bar() -> None:
+    def inner() -> str: return "bar"
+    functools.update_wrapper(inner, bar)  # type: ignore
+    print(inner.__dict__)  # type: ignore
```
[the remark this answers is not in the document]
Comment (hauntsaninja) [reply mypyc/test-data/run-functions.test:1284]: I think it's okay as is, I kinda feel obfuscation here makes it less clear to the reader what is being tested

[the remark this answers is not in the document]
Comment (AlexWaygood) [reply mypyc/test-data/run-functions.test:1284]: yeah i didn't really like this either, was just throwing it out there as an option :)

**Comment by hauntsaninja** [comment]
Merging since this has been open for a bit and unblocks Python 3.12 mypyc in black, let me know if this change is actually incorrect!

**Pull request merged**